File: src/onedimensiondp/lc_139_word_break/word_break.py
class Solution:
    def wordBreak(self, s: str, wordDict: list[str]) -> bool:
        wordDict = set(wordDict)
        memo = {}
        
        def dp(s, res):
            if s in memo:
                return memo[s]
            if s == '' or s in wordDict:
                return True

            for i in range(len(s)):
                res += s[i]
                if res in wordDict:
                    memo[s] = dp(s[i+1:], '')
                    if memo[s]:
                        return True
            
            return False
        return dp(s, '')

File: src/onedimensiondp/lc_139_word_break/word_break.java
import java.util.*;

class Solution {

    Set<String> wordSet;
    Map<String, Boolean> memo = new HashMap<>();

    public boolean wordBreak(String s, List<String> wordDict) {
        wordSet = new HashSet<>(wordDict);
        return dp(s, "");
    }

    public boolean dp(String s, String build) {
        if (memo.containsKey(s))
            return memo.get(s);
        if (wordSet.contains(s) || s.equals(""))
            return true;
        for(int i = 0; i < s.length(); i++){
            build = build + s.charAt(i);
            if (wordSet.contains(build)) {
                memo.put(s, dp(s.substring(i+1), ""));
                if (memo.get(s))
                    return true;
            }
        }
        return false;
    }
}